[Infrastructure] Implement custom dropdown widget (Resolves #1527, Resolves #731, Resolves #689)

angelosilvestre opened the pull request:
[Infrastructure] Implement custom dropdown widget. Resolves #1527, Resolves #731, Resolves #689

Tapping at any `DropdownButton` in the example app toolbar causes the toolbar to immediately close. The issue is that Flutter `DropdownButton` displays the dropdown in its own route, which steals the focus from the editor.

This PR introduces a custom dropdown button. I called it `SuperDropdownButton`, but we can choose a better name. This widget displays the dropdown using an `OverlayPortal`. By doing so, it can share focus with `SuperEditor`.

We need to define how we want the public API:

- What we want to make everything.
- What we should do by default.

For example: What should be its default height? Do we want to make the width as large as the largest item in the list or just enough to fit the selected item? 

This is how it looks so far:

https://github.com/superlistapp/super_editor/assets/7597082/d2af4fd7-f60a-4eef-9ed6-a71cfee6ed15

I noticed that changing the paragraph alignment doesn't immediately rebuilds the paragraph, but it isn't related to the dropdown. I will file a ticket after this PR is merged.

Another example, constraining the dropdown height:

https://github.com/superlistapp/super_editor/assets/7597082/ff45cd8c-5666-47a5-99c9-60e21f2c505e

matthew-carroll commented:
@angelosilvestre can we position the popover below the button like a more traditional popup menu? Also can we round the corners to fit better with the toolbar shape?

Generally the functionality looks good. Let's make it look the best we can for the Super Editor Example app, and then we can look at which aspects should be configurable.

matthew-carroll commented on super_editor/example/lib/demos/example_editor/_toolbar.dart:
```diff
@@ -224,6 +224,9 @@ class _EditorToolbarState extends State<EditorToolbar> {
     }
+
+    // Rebuild to display the selected item.
+    setState(() {});
```
Where is the selected item stored? Why do we want `setState()` to be empty?

angelosilvestre replied:
The dropdown itself doesn't store the selected value, it's always provided via constructor.

In the toolbar, this value also isn't stored, we query the node type from the selected node when we create the dropdown.

matthew-carroll replied:
I don't understand those explanations. If the selected item isn't stored, then why are calling `setState()`?

angelosilvestre replied:
I'm calling `setState` to rebuild the toolbar. Then, it will query the value again.

angelosilvestre replied:
After taking a better look at this, I think that the main problem here is that we aren't rebuilding the toolbar when the document changes.

After we change the selected node's metadata, we need to rebuild the toolbar so it can "see" the current state of the selected node.

I moved this `setState` call to the document change listener. Please let me know if this is an acceptable solution.

matthew-carroll commented on super_editor/example/lib/demos/example_editor/_toolbar.dart:
```diff
@@ -521,26 +527,30 @@ class _EditorToolbarState extends State<EditorToolbar> {
+                        ),
+                      ),
+                    ),
+                    parentFocusNode: widget.editorFocusNode,
```
Let's not hide properties with simple values down below builders. They're too easy to miss.

matthew-carroll commented on super_editor/example/lib/demos/example_editor/_toolbar.dart:
```diff
@@ -584,26 +594,14 @@ class _EditorToolbarState extends State<EditorToolbar> {
+                    items: [TextAlign.left, TextAlign.center, TextAlign.right, TextAlign.justify],
                     onChanged: _changeAlignment,
+                    boundaryKey: widget.editorViewportKey,
+                    parentFocusNode: widget.editorFocusNode,
```
Please move `parentFocusNode` to top of the property list and then `boundaryKey` right below that.

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,479 @@
+    this.dropdownKey,
+  });
+
+  /// The items that will be displayed in the dropdown list.
```
Let's re-order these properties to match our typical ordering.

angelosilvestre replied:
Re-ordered to match how we are using it on `_toolbar.dart`

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,479 @@
+  final ScrollController _scrollController = ScrollController();
+
+  @visibleForTesting
+  int? get focusedIndex => _focusedIndex;
```
Let's put focus up top above the controllers

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,479 @@
+/// when the dropdown is visible.
+///
+/// This widget doesn't enforce any style, dropdown position or decoration.
+class RawDropdown extends StatefulWidget {
```
What does "Raw" mean here?

angelosilvestre replied:
It only mean that we don't make any decisions here about how the widget should look like. We could probably find a better name for it... `UnstyledDropdown`, `OverlayDropdown`,  or something else.

matthew-carroll replied:
Can you elaborate on that explanation? You've said what this widget "isn't" doing. Can you describe what it "is" doing?

angelosilvestre replied:
I tried to summarize what this widget does in the previous lines:

- It listens for changes of its `OverlayPortalController` and shows/hides the `Overlay`.
- Setup focus sharing.
- Requests focus to the dropdown's `FocusNode` when the overlay is displayed.
- Delivers key events to `onKeyEvent`. 
- Closes the dropdown when tapping outside of it.

matthew-carroll replied:
I'm confused why the follower and aligner are configured elsewhere instead of here? If this widget does what you describe, why wouldn't it also include positioning?

Is this widget expected to be used publicy? If not, can we make it private?

If we make it private and move the follower and aligner in here, let's call this `_FollowerPopover`. The content of the popover can then be provided via `child` to decide what to display within the popover.

angelosilvestre replied:
I've made it that way so we (or others) could re-use the `RawDropDown` to implement other kinds of dropdown buttons.

For example, it would be relatively easy to implement something like a simple (and incomplete) color picker with the following code:

```dart
class ColorPicker extends StatefulWidget {
  const ColorPicker({super.key});

  @override
  State<ColorPicker> createState() => _ColorPickerState();
}

class _ColorPickerState extends State<ColorPicker> {
  final DropdownController _dropdownController = DropdownController();

  // This might be removed.
  final FocusNode _focusNode = FocusNode();

  final List<MaterialColor> colors = [
    Colors.red,
    Colors.blue,
    Colors.amber,
    Colors.brown,
    Colors.green,
    Colors.orange,
    Colors.pink,
    Colors.teal
  ];

  @override
  void dispose() {
    _dropdownController.dispose();
    _focusNode.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return RawDropdown(
      controller: _dropdownController,
      dropdownBuilder: _buildColorPopover,
      parentFocusNode: _focusNode,
      child: ElevatedButton(
        onPressed: () => _dropdownController.open(),
        child: Text("Pick a color"),
      ),
    );
  }

  Widget _buildColorPopover(BuildContext context, LeaderLink link, FollowerBoundary boundary) {
    return Follower.withOffset(
      link: link,
      leaderAnchor: Alignment.bottomCenter,
      followerAnchor: Alignment.topCenter,
      offset: Offset(0, 10),
      child: Material(
        elevation: 4.0,
        borderRadius: BorderRadius.circular(8.0),
        child: Container(
          height: 50,
          width: 200,
          padding: const EdgeInsets.all(8.0),
          child: ListView.separated(
            scrollDirection: Axis.horizontal,
            // TODO: implement selection
            itemBuilder: (context, index) => Container(
              decoration: BoxDecoration(
                shape: BoxShape.circle,
                color: colors[index],
              ),
              height: 20,
              width: 20,
            ),
            separatorBuilder: (context, index) => SizedBox(width: 10),
            itemCount: colors.length,
          ),
        ),
      ),
    );
  }
}
```

https://github.com/superlistapp/super_editor/assets/7597082/3cecd23f-8bbb-46c0-ae26-8f8503430363

matthew-carroll commented on super_editor/test/infrastructure/dropdown_test.dart:
```diff
@@ -0,0 +1,470 @@
+import 'package:super_editor/super_editor_test.dart';
+
+import '../super_editor/test_documents.dart';
+
```
Let's call this file `super_dropdown_test` to make it clear that we're testing our custom widget instead of some arbitrary dropdown.

angelosilvestre commented:
> @angelosilvestre can we position the popover below the button like a more traditional popup menu? Also can we round the corners to fit better with the toolbar shape?

@matthew-carroll  Yes, initially I left the dropdown centered to match the material dropdown.

Updated:

https://github.com/superlistapp/super_editor/assets/7597082/3b7e37d8-e8f5-4a8a-a350-84d8e01413b8

With slow animations:

https://github.com/superlistapp/super_editor/assets/7597082/db9f3a57-8393-4dca-bdf4-9a364568c785

matthew-carroll commented:
@angelosilvestre I see in your latest video that the popover prevents itself from moving beyond the bottom of the screen, but that causes it to cover the toolbar. My first thought is that this is unexpected and we don't want it. Was there any particular reason that you went with a screen-constrained popup? 

The following heuristic comes to mind:
 * When there's enough room beneath the toolbar, show the popup below the toolbar, otherwise
 * When there's enough room above the toolbar, show the popup above the toolbar, otherwise
 * If there's a reasonable amount of space (enough to see a few items), make the popover shorter and re-apply above rules, otherwise
 * Pin the popup in the standard position and size below the button and let it go off-screen

angelosilvestre commented:
@matthew-carroll As discussed, I implemented a positioning behavior similar to Notion's.

1. If there is enough room to display the dropdown list beneath the toolbar, position it below the toolbar.
2. If there is enough room to display the dropdown list above the toolbar, position it above the toolbar.
3. Pin the dropdown list to the bottom and let it cover the toolbar.

I also simplified the transition to be just a fade-in animation.

This is how it's looking right now:

https://github.com/superlistapp/super_editor/assets/7597082/06a9adaa-84ea-4346-b200-d0a43030724d

For reference, this is how Notion's dropdown looks:

https://github.com/superlistapp/super_editor/assets/7597082/134118e7-0163-43c0-ac03-666b5be1008a

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,506 @@
+import 'package:follow_the_leader/follow_the_leader.dart';
+import 'package:super_editor/super_editor.dart';
+
+/// A button which displays a dropdown with a vertical list of items.
```
Let's replace all uses of "dropdown" with "popover list" or "dropdown popover" so that it remains clear the dropdown is a separate piece of UI which sits above the selected item.

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,506 @@
+/// A button which displays a dropdown with a vertical list of items.
+///
+/// Unlike Flutter `DropdownButton`, which displays the dropdown in a separate route,
+/// this widget displays its dropdown in an `Overlay`. By using an `Overlay`, focus can shared
```
"focus can shared" -> "focus can be shared"

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,506 @@
+/// with the [parentFocusNode]. This means that when the dropdown requests focus, [parentFocusNode]
+/// still has non-primary focus.
+///
+/// The dropdown tries to fit all items on the available space. If there isn't enough room,
```
Let's simply enumerate the rules for clarity:

```
/// The popover list is positioned based on the following rules:
///
///    1. The popover is displayed below the selected item, if there's enough room, or
///    2. The popover is displayed above the selected item, if there's enough room, or
///    3. The popover is displayed with its bottom aligned with the bottom of
///         the given boundary, and it covers the selected item.
```

matthew-carroll replied:
I didn't mention scrolling constraints in there. Please update accordingly. The goal is to make the rules quick and easy to understand.

angelosilvestre replied:
I'm not sure I understood what "scrolling constraints" means here.

matthew-carroll replied:
I'm not entirely sure what I meant, but I think I may have been referring to the content in the popover going from fully-visible to partially visible + scrolling. For example, let's say we pull up a list of options in a popover. There's likely some kind of decision whether to make the popover as tall as all those items in the list, or whether to make the popover shorter, and introduce scrolling to the list. That choice about adding scrolling, or not, and the resulting height, impacts all of the rules above. They're like two sides of the same coin.

angelosilvestre replied:
Updated.

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,506 @@
+///
+/// The dropdown tries to fit all items on the available space. If there isn't enough room,
+/// the list of items becomes scrollable with an always visible toolbar.
+/// Provide [dropdownContraints] to enforce aditional constraints on the dropdown list.
```
Let's not allude to what `dropdownConstraints` is for. It's not clear what it means to "enforce additional constraints". Please provide the specific motivation and behavior for a developer to provide `dropdownConstraints`.

angelosilvestre replied:
Updated.

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,506 @@
+/// the list of items becomes scrollable with an always visible toolbar.
+/// Provide [dropdownContraints] to enforce aditional constraints on the dropdown list.
+///
+/// The user can navigate between the options using the arrow keys, select an option with ENTER and
```
Rather than passively mention interactions, please specify precise impact.

```
/// The popover list includes keyboard selection behaviors:
/// 
///   * Pressing UP/DOWN moves the "active" item selection up/down.
///   * Pressing UP with the first item active, or DOWN with the last item active, does nothing.
///   * Pressing ENTER selects the currently active item and closes the popover list.
```

angelosilvestre replied:
Just a note, pressing UP with the first item active moves the active selection to the last item and  pressing DOWN with the last item active moves the active selection to the first item.

matthew-carroll replied:
Is that the most common behavior when at the top or bottom of a list with arrow keys?

angelosilvestre replied:
From what I've seen, it's pretty common.

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,506 @@
+///
+/// The user can navigate between the options using the arrow keys, select an option with ENTER and
+/// close the dropdown with ESC.
+class SuperDropdownButton<T> extends StatefulWidget {
```
Let's find a better name for this. What about `ItemSelector`? Something else?

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,506 @@
+    this.dropdownKey,
+  });
+
+  /// [FocusNode] which will share focus with the dropdown list.
```
Please replace with:

```
/// The [FocusNode], to which the popover list's [FocusNode] will be added as a child.
///
/// In Flutter, [FocusNode]s have parents and children. This relationship allows an
/// entire ancestor path to "have focus", but only the lowest level descendant
/// in that path has "primary focus". This path is important because various
/// widgets alter their presentation or behavior based on whether or not they
/// currently have focus, even if they only have "non-primary focus".
///
/// When the popover list of items is visible, that list will have primary focus.
/// Moreover, because the popover list is built in an `Overlay`, none of your
/// widgets are in the natural focus path for that popover list. Therefore, if you
/// need your widget tree to retain focus while the popover list is visible, then
/// you need to provide the [FocusNode] that the popover list should use as its
/// parent, thereby retaining focus for your widgets.
```

matthew-carroll replied:
Should the parent focus node be required? What if a developer wants to use this widget in a place where he doesn't care about sharing focus?

angelosilvestre replied:
The dropdown internally uses a `SuperEditorPopover`. Should we pass a `FocusNode` created inside the dropdown if the app doesn't provide one?

matthew-carroll replied:
Is `SuperEditorPopover` something we created in this PR or did it already exist? If we created it in this PR, then can we just not make it required there either? If that already existed and we don't want to change it, then yes, I think create a new `FocusNode()` locally and pass it through.

angelosilvestre replied:
It was introduced in https://github.com/superlistapp/super_editor/pull/1443. It was created specifically to share focus, so I think we don't want to change it.

I'm creating a local `FocusNode`.

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,506 @@
+  /// [FocusNode] which will share focus with the dropdown list.
+  final FocusNode parentFocusNode;
+
+  /// A [GlobalKey] to a widget that determines the bounds where the dropdown can be displayed.
```
Should this be required? Can we default to a screen boundary if its `null`?

angelosilvestre replied:
Updated.

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,506 @@
+
+  /// A [GlobalKey] to a widget that determines the bounds where the dropdown can be displayed.
+  ///
+  /// Used to avoid the dropdown to be displayed off-screen.
```
Please try to be more specific. The person reading probably has no idea how this is "used to avoid the dropdown to be displayed off-screen".

angelosilvestre replied:
Updated.

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,506 @@
+  /// Used to avoid the dropdown to be displayed off-screen.
+  final GlobalKey boundaryKey;
+
+  /// The currently selected value or `null` if no item is selected.
```
Please make clear the connection between this value and the widget's display.

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,506 @@
+  /// The currently selected value or `null` if no item is selected.
+  final T? value;
+
+  /// The items that will be displayed in the dropdown list.
```
Same here. There's ambiguity between the concept of the value and the items vs the display of the value and the display of the items.

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,506 @@
+  /// The items that will be displayed in the dropdown list.
+  final List<T> items;
+
+  /// Called when the user selects an item on the dropdown list.
```
Please enumerate the ways in which a user might "select an item in the popover list"

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,506 @@
+  /// Called when the user selects an item on the dropdown list.
+  final ValueChanged<T?> onChanged;
+
+  /// The background color of the focused list item.
```
There are few issues with this property.

First, this level of control is too narrow. What if the user wants to use a gradient? Or a border? Or a sparkle effect? We shouldn't be making this UI decision for the user.

Second, it's not obvious what it means for an item to "have focus". In fact, I don't think the item will have focus, will it? Isn't it the list that has focus?

I think what you're trying to describe here is usually called the "active" item. You can double check that. But I know the terms "selected" and "active" often complement one another.

Wherever we want to refer to the item that's selected in the list, but not the actual selected item, we should introduce an appropriate term and use is consistently through this documentation.

angelosilvestre replied:
I changed to use a `BoxDecoration`, if you want we should still offer more control, then we could use an widget builder.

About the naming, I updated to use the term "active" to represent the item selected in the list.

matthew-carroll replied:
Can you let me know why we ask the user to specify anything at all? Whether a color or a decoration? Do we really need this property? Can the user not achieve the same result through widgets that the user is already building?

angelosilvestre replied:
To remove that property, we need to add another parameter in `itemBuilder` to indicate whether the item is active. Should we go in this direction?

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,506 @@
+  /// The background color of the focused list item.
+  final Color? focusColor;
+
+  /// A [GlobalKey] bound to the dropdown list.
```
Please describe why someone cares. Why is this provided? What does it mean if it's `null`? Etc.

angelosilvestre replied:
Currently, I'm only using this key to find the popover list during tests. Should I remove this and find the popover list with another approach?

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,506 @@
+  /// A [GlobalKey] bound to the dropdown list.
+  final GlobalKey? dropdownKey;
+
+  /// Constraints applied to the dropdown list.
```
Please describe HOW this constrains the popover list.

angelosilvestre replied:
I'm not sure how I should describe that... The only thing I'm doing with this `BoxConstraints` is passing it to a `ConstrainedBox` which wraps the popover list.

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,506 @@
+  /// Constraints applied to the dropdown list.
+  final BoxConstraints? dropdownContraints;
+
+  /// Builds each item in the dropdown list.
```
Please describe the connection between this builder and the items property.

Also, is there not already a typedef for a builder like this?

angelosilvestre replied:
I didn't find any typedef for this.

matthew-carroll replied:
Let's typedef it ourselves then

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,506 @@
+      return KeyEventResult.ignored;
+    }
+
+    if (![
```
Can you declare this list `const` so we're not re-instantiating it on every key event? Also, if order doesn't matter, then this should be a `Set` not a `List`.

angelosilvestre replied:
We can't declared it as a `const Set` because of the following compilation error:

```console
An element in a constant set can't override the '==' operator, or 'hashCode', but the type 'LogicalKeyboardKey' does.
Try using a different value for the element, or removing the keyword 'const' from the set.
```

Should we use a non-const `Set` or a const `List`?

matthew-carroll replied:
Let's use a `const List()`

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,506 @@
+    }
+
+    if (event.logicalKey == LogicalKeyboardKey.escape) {
+      _dropdownController.hide();
```
`hide()` sounds temporarily. Let's call this `close()` unless you have a reason not to.

matthew-carroll replied:
Ok, we can try that

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,506 @@
+    }
+
+    if (event.logicalKey == LogicalKeyboardKey.enter || event.logicalKey == LogicalKeyboardKey.numpadEnter) {
+      if (_focusedIndex == null) {
```
Please describe what this condition means with a comment.

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,506 @@
+        return KeyEventResult.handled;
+      }
+
+      _submitItem(widget.items[_focusedIndex!]);
```
Let's call this `_selectItem()` to remain consistent with all other places where we talk about this widget displaying a selected item.

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,506 @@
+    int? newFocusedIndex;
+    if (event.logicalKey == LogicalKeyboardKey.arrowDown) {
+      if (_focusedIndex == null || _focusedIndex! >= widget.items.length - 1) {
+        // We don't have a focused item or we are at the end of the list. Focus the first item.
```
What does it mean to not have a focused item? Is that not an error state?

angelosilvestre replied:
In the current implementation, the popover list opens without an "active item". The active item is set only after the user presses ARROW DOWN or ARROW UP.

This is compatible with other popover list implementations.

matthew-carroll replied:
Really? If we're looking at a typical address form on the web, and there's a country selector, and we have United States selected, when we open that selector and show the popover, "United States" isn't active? And the popover isn't centered on the "United States" item?

On a related note, if no item is active initially, which one becomes active when you press UP or DOWN?

angelosilvestre replied:
> Really? If we're looking at a typical address form on the web, and there's a country selector, and we have United States selected, when we open that selector and show the popover, "United States" isn't active? And the popover isn't centered on the "United States" item?

It makes sense. Updated to make the selected value "active" in the list.

> On a related note, if no item is active initially, which one becomes active when you press UP or DOWN?

Pressing UP makes the last item active and pressing DOWN makes the first item active. We could also change the behavior to make the first item active if there isn't a selected item.

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,506 @@
+      parentFocusNode: widget.parentFocusNode,
+      onKeyEvent: _onKeyEvent,
+      child: ConstrainedBox(
+        // TODO: what value should we use?
```
When I first read this, I thought this was the popover list constraint. But this is constraining the selected item display, right? If so, why are we constraining that at all?

angelosilvestre replied:
I don't remember why I added that, but it seems it's unnecessary. Removed.

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,506 @@
+  }
+
+  Widget _buildSelectedItem() {
+    return Row(
```
Should this be a `Row` or a `Stack`? Imagine the developer wants to display a selected item with their own background color or gradient, how would they get a full-bleed display if we're pushing their widget to the side for a dropdown arrow?

matthew-carroll commented on super_editor/example/lib/demos/example_editor/_toolbar.dart:
```diff
@@ -95,14 +97,32 @@ class _EditorToolbarState extends State<EditorToolbar> {
+    // The document has changed.
+    // Some of the selected node's metadata, for example, the blockType, influences how the toolbar is displayed.
+    // Reflow the toolbar to acount for the new state of the selected node.
+    setState(() {});
```
Is there any particular reason that we don't listen for this change around, or within the pieces of the toolbar that care, rather than rebuilding the whole toolbar?

angelosilvestre replied:
Updated.

matthew-carroll commented on super_editor/example/lib/demos/example_editor/_toolbar.dart:
```diff
@@ -521,26 +541,30 @@ class _EditorToolbarState extends State<EditorToolbar> {
                 Tooltip(
                   message: AppLocalizations.of(context)!.labelTextBlockType,
-                  child: DropdownButton<_TextType>(
+                  child: ItemSelector<_TextType>(
```
I'm wondering if there's a more useful way to breakdown responsibilities for these widgets. After you go through the other new comments, let's jump on a call and talk about how to break things up.

My gut feeling is that we want two different levels of widgets here. We want foundational pieces that make almost zero visual designs but provide some base level relationships. Then we want some widgets that give us default styling, and also take a few properties to configure obvious UI details. At the moment, we're trying to do both of those things in one group of widgets.

angelosilvestre replied:
It makes sense, we can jump on a call to talk about that.

rutvik110 commented:
> I noticed that changing the paragraph alignment doesn't immediately rebuilds the paragraph, but it isn't related to the dropdown. I will file a ticket after this PR is merged.

@angelosilvestre I'm taking look at it as part of the #1407

angelosilvestre commented:
@matthew-carroll As discussed, the API is changing. Some things are a bit different from what we've talked. Here's the summary:

We have now the following widgets:

### `PopoverScaffold` (still needs a better name), which takes: 

- A `PopoverController` that opens and closes the popover.
- A `buttonBuilder` to build the button.
- A `popoverBuilder` to build the popover.
- A `popoverGeometry` to control the size and position of the popover.
- An `onKeyEvent` to which key events will be forwarded. This widget uses a `SuperEditorPopover` to setup the focus sharing, and `SuperEditorPopover` is the widget which install the `Focus` widget on the tree. Because of that, instead of adding a `Focus` ourselves, we just forward the `onKeyEvent`.
- An optional `parentFocusNode` to share focus
- An optional `boundaryKey` to keep the popover within the boundary.

This widget doesn't take a list of items because it doesn't care about whats being displayed on the popover. We could use the popover to display a text, an image preview, etc.

### `ItemSelectionList`, which takes:
- The selected value.
- The list of items.
- A `buttonBuilder` to build the button.
- An `itemBuilder` to build the each item on the list.
- An  `onItemActivated` to notify when an item is activated.
- An `onItemSelected` to notify when an item is selected.
- A `popoverGeometry` to control the size and position of the popover.
- An optional `parentFocusNode` to share focus.
- An optional `boundaryKey` to keep the popover within the boundary.

This widget handles the keyboard navigation, auto-scrolling and builds the list of options. I ignored popovers with horizontal scrolling for now. I'm not sure if we need a widget in between  `PopoverScaffold` and `ItemSelectionList`, which instead of an `itemBuilder` would take a `popoverBuilder` to build the entire popover based on the list of `items`.

### `SuperEditorDemoTextItemSelector` and  `SuperEditorDemoIconItemSelector`:

Widgets with a predefined style to be used in the demos.

angelosilvestre commented:
Rebased to fix an import.

matthew-carroll commented:
@angelosilvestre can you please edit your comment up above to point out what you changed from our discussion and why? 

Based on what I'm seeing in that comment right now, here are a couple comments:

## PopoverScaffold
> An onKeyEvent to which key events will be forwarded. This widget uses a SuperEditorPopover to setup the focus sharing, and SuperEditorPopover is the widget which install the Focus widget on the tree. Because of that, instead of adding a Focus ourselves, we just forward the onKeyEvent.

What do you mean "This widget uses a SuperEditorPopover"? The scaffold is the lowest level widget here and `SuperEditorPopover` is the highest level. If the scaffold is truly using a `SuperEditorPopover`, then that needs to be reversed.

"instead of adding a Focus ourselves" - I'm not sure what that means either, but you said the scaffold takes a `parentFocusNode`, which indicates that the scaffold knows about and cares about focus. Unless I'm missing something, we don't want to split focus responsibilities across multiple widgets.

## ItemSelectionList
Why is this list taking a geometry? I believe the purpose of this widget, as discussed on our call, was to provide a list widget. That list widget might be placed within a popover, but its not, itself, a popover. It's a list that understands the concept of "selection" and "activation" and arrow key navigation, and ENTER to select the activated item, and auto-scrolling when pressing arrows keys near the top and bottom of the visible space. All of that is independent from use within a popover. Such a widget would be equally applicable for a selection list that's placed within a typical screen UI.

Please let me know if I'm missing something.

angelosilvestre commented:
> What do you mean "This widget uses a SuperEditorPopover"? The scaffold is the lowest level widget here and SuperEditorPopover is the highest level. If the scaffold is truly using a SuperEditorPopover, then that needs to be reversed.

Maybe we have some naming issue... `SuperEditorPopover` is a widget introduce in another PR. It sets up the focus sharing and also add the blocked intents needed for macOS. It's this widget:

https://github.com/superlistapp/super_editor/blob/6e2040f573e58fd6e88224fce66f3719fe30b0b7/super_editor/lib/src/infrastructure/popovers.dart#L5-L64

> Why is this list taking a geometry? I believe the purpose of this widget, as discussed on our call, was to provide a list widget. That list widget might be placed within a popover, but its not, itself, a popover. It's a list that understands the concept of "selection" and "activation" and arrow key navigation, and ENTER to select the activated item, and auto-scrolling when pressing arrows keys near the top and bottom of the visible space. All of that is independent from use within a popover. Such a widget would be equally applicable for a selection list that's placed within a typical screen UI.

I think I misunderstood it a little bit. I'll update it.

angelosilvestre commented:
@matthew-carroll  Updated the PR. I think now it's closer to what we've talked.

I temporarily commented the test until we are sure the API is right.

Can you do a first pass review?

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,1059 @@
+///   * Pressing UP with the first item active moves the active item selection to the last item.
+///   * Pressing DOWN with the last item active moves the active item selection to the first item.
+///   * Pressing ENTER selects the currently active item and closes the popover list.
+class SuperEditorDemoTextItemSelector extends StatefulWidget {
```
Given that this is a demo-oriented class, it should be defined within the example app project. Only the core popover building blocks should be defined within the package and exported.

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,1059 @@
+    this.value,
+    required this.items,
+    required this.onSelected,
+    this.parentFocusNode,
```
Please put focus nodes near the top of property lists. We typically place generic Flutter details first, e.g., `Key`, `BuildContext`, `FocusNode`, controllers.

Also, please move `boundaryKey` up as well, so that it's above `items`.

And, in general, ensure that callbacks tend to be the last thing in property lists, other than `child` or `children` parameters.

Make similar changes to the actual property declarations.

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,1059 @@
+          onItemSelected: _onItemSelected,
+          onCancel: () => _popoverController.close(),
+          focusNode: _popoverFocusNode,
+          parentFocusNode: widget.parentFocusNode,
```
In the comments, we discussed that `ItemSelectionList` should only include list-related behaviors. Does it really need the `parentFocusNode`? Could `PopoverScaffold` setup the focus relationship and have the `ItemSelectionList` be unaware of that parenting relationship?

angelosilvestre replied:
If we don't pass the `parentFocusNode` to the list, we need a way to setup the key event handler in `PopoverScaffold` and find a way to forward this key events to `ItemSelectionList`.

matthew-carroll replied:
Why exactly do we need the list to setup the parent focus relationship? Why can't the scaffold setup the parent relationship and let the list operate like any other list with focus?

angelosilvestre replied:
We can setup the parent focus relationship in the scaffold with `FocusWithCustomParent`. However, in `ItemSelectionList` with have a `Focus` widget to setup the key handler. If we just declare it like the following:

```dart
// ItemSelectionListState

@override
Widget build(BuildContext context) {
  return Focus(
     focusNode: widget.focusNode,
     onKeyEvent: _onKeyEvent,
     ...
  );
}
```

then the focus sharing doesn't work anymore. When `PopoverScaffold` requests focus, the `widget.focusNode` loses focus.

Am I missing something here?

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,1059 @@
+        onTap: onTap,
+        child: Container(
+          constraints: const BoxConstraints(minHeight: kMinInteractiveDimension),
+          alignment: AlignmentDirectional.centerStart,
```
What's this `AlignmentDirectional` about? I don't think I've seen that before.

angelosilvestre replied:
Is just an `Alignment` that is directionality aware, so `centerStart` is `centerLeft` when left-to-right and `centerRight` when right-to-left, but I guess we could just use `Alignment`.

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,1059 @@
+  }
+
+  Widget _buildButton(BuildContext context) {
+    return PopoverArrowButton(
```
What's the relevance of saying "arrow" in this name? Did we have that term in there on our call? I thought I remembered two relevant variations: a text button and an icon button. Also, given that this is a styled button, it should probably be named as a Super Editor demo and moved into the Example app, right?

angelosilvestre replied:
This widget is used by both the text button and the icon button. I moved this widget to the demo app.

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,1059 @@
+  }
+}
+
+/// A rounded rectangle shape with a fade-in transition.
```
I don't think we should combine the shape with the transition animation. Those don't seem related. Also, this popover shape doesn't allow any configuration for the shape...

angelosilvestre replied:
I'm currently using a `Material` widget to build the popover shape, should I take a `ShapeBorder` to customize the popover shape?

matthew-carroll replied:
I'm confused - how does the use of `Material` relate to the two issues I raised above?

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,1059 @@
+  }
+}
+
+/// A widget which displays a button built in [buttonBuilder] with a popover
```
Imagine reading this first sentence in docs - would you have any idea what this widget is meant to help with? Rather than describe this widget with a very generic concept of buttons and following, please describe this widget in terms of the UX that the user is looking for, i.e., a popover selection system.

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,1059 @@
+  }
+}
+
+/// Controls the size and position of a popover.
```
This data structure doesn't control anything. This data structure *is* the offset and size of a popover in screen space. Please try to avoid the term "position" when you're not talking about text or document positions.

angelosilvestre replied:
Updated.

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,1059 @@
+}
+
+/// A [FollowerAligner] which uses a [delegate] to align a [Follower].
+class _DelegateAligner implements FollowerAligner {
```
Does this mean we need a functional implementation of `FollowerAligner`? Or perhaps a default definition of `FollowerAligner` that takes a delegate? If so, let's make that change in follow_the_leader.

angelosilvestre replied:
OK, should I open a PR adding a `FunctionalAligner` to follow_the_leader?

angelosilvestre replied:
Opened https://github.com/Flutter-Bounty-Hunters/follow_the_leader/pull/42

angelosilvestre replied:
Updated to use the `FunctionalAligner`. I temporarily added a dependency override to use `follow_the_leader` from git. We'll need to release a new version...

matthew-carroll replied:
I just released a new version

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,1059 @@
+  return alignment;
+}
+
+/// A function to align a Widget following a leader Widget.
```
Here again, similar to the description of the popover scaffold, you seem to be trying to avoid saying what this thing is actually meant to do. Why would we want to describe this as "A function to align a widget following a leader widget"? Also, doesn't every follower aligner fit that definition? Let's tell the reader what they're goin to want to know about when/where/why/how to use this thing.

angelosilvestre replied:
Now that https://github.com/Flutter-Bounty-Hunters/follow_the_leader/pull/42 was merged, this won't be used anymore.

matthew-carroll reviewed (commented):
I did an initial review. One big initial change should be to move all the demo stuff into the demo app, instead of in the package. That might clarify some of the relationships and also point out where some things are named poorly.

angelosilvestre commented on super_editor/example/lib/demos/example_editor/_toolbar.dart:
```diff
@@ -521,26 +522,33 @@ class _EditorToolbarState extends State<EditorToolbar> {
-                              child: Padding(
-                                padding: const EdgeInsets.only(left: 16.0),
-                                child: Text(_getTextTypeName(textType)),
+                  child: _DocumentListenableBuilder(
```
With the recent changes we don't need it anymore.

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,635 @@
+///
+/// When the popover is displayed it requests focus to itself, so the user can
+/// interact with the content using the keyboard.
+class PopoverScaffold extends StatefulWidget {
```
Let's move this scaffold to the top of the file because everything else is pretty much in service of this.

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,635 @@
+  }
+}
+
+/// A widget used to build UI controls that display a following popover.
```
Let's replace the top line with this: "A scaffold, which builds a popover selection system, comprised of a button and a popover that's positioned near the button."

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,635 @@
+  final FocusOnKeyEventCallback? onKeyEvent;
+
+  /// [FocusNode] which will share focus with the popover.
+  final FocusNode? popoverFocusNode;
```
Let's move this up beneath `popoverGeometry` to keep `popover...` things together.

Also, can you remind me whey we need this? The dev builds the popover, right?

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,635 @@
+  final PopoverGeometry popoverGeometry;
+
+  /// Called at each key press while the popover has focus.
+  final FocusOnKeyEventCallback? onKeyEvent;
```
Can you remind me why we need to let developers provide an `onKeyEvent`? If devs want to process keys within the button, they can build that into the button, and if devs want to process keys within the popover list, they can build that into the list, right?

angelosilvestre replied:
It seems it isn't being used anymore. Removed.

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,635 @@
+        devicePixelRatio: MediaQuery.devicePixelRatioOf(context),
+      );
+    } else {
+      _screenBoundary = ScreenFollowerBoundary(
```
Given that we're installing a screen boundary when the boundary key is null, please revisit all the Dart Docs that discuss the boundary key, because I think they say the popover might go off-screen.

angelosilvestre commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,635 @@
+
+  final GlobalKey _scrollableKey = GlobalKey();
+
+  /// Holds keys to each item on the list.
```
Updated

angelosilvestre commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,635 @@
+    }
+  }
+
+  /// Called when the user taps an item or presses ENTER with an active item.
```
I ended up removing this method and calling `onItemActivated` directly.

angelosilvestre commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,635 @@
+    widget.onItemSelected(item);
+  }
+
+  /// Activates the item at [itemIndex] and ensure it's visible on screen.
```
Updated.

angelosilvestre commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,635 @@
+      widget.onItemActivated?.call(widget.items[itemIndex]);
+    }
+
+    // Scrolls on the next frame to let the popover to be
```
Added a comment explaining.

angelosilvestre commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,635 @@
+
+      return KeyEventResult.handled;
+    }
+
```
Updated.

angelosilvestre commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,635 @@
+import 'package:super_editor/src/infrastructure/platforms/mac/mac_ime.dart';
+import 'package:super_editor/super_editor.dart';
+
+/// A selection control, which displays a selected item, and upon tap, displays a
```
Updated the doc. I could use some help to add an information in the doc about the distinction between "active" and "selected"...

matthew-carroll commented on super_editor/lib/src/infrastructure/dropdown.dart:
```diff
@@ -0,0 +1,608 @@
+
+/// A list where the user can navigate between its items and select one of them.
+///
+/// This widget shares focus with its [parentFocusNode]. This means that when the list requests focus,
```
As discussed, let's get to the bottom of why this list needs the parent focus node.  It seems like the following should work:

Editor focus -> popover focus (using parent relationship) -> list focus (using regular focus relationship)

Let's find out why that isn't working.

angelosilvestre replied:
I pushed a commit with this change. Now it's working as expected.

angelosilvestre commented:
@matthew-carroll Added tests

matthew-carroll commented on super_editor/lib/src/infrastructure/selectable_list.dart:
```diff
@@ -0,0 +1,268 @@
+  }
+}
+
+/// Builds a popover list item.
```
Should "popover" be mentioned here?

matthew-carroll commented on super_editor/lib/src/infrastructure/selectable_list.dart:
```diff
@@ -0,0 +1,268 @@
+
+/// Builds a popover list item.
+///
+/// [isActive] is `true` if [item] is the currently active item on the list, or `false` otherwise.
```
It's probably a good idea to say something about what it means to be active, e.g., the item in the list that currently has focus and can be moved up/down with arrow keys.

angelosilvestre replied:
Updated.

matthew-carroll commented on super_editor/test/infrastructure/popover_test.dart:
```diff
@@ -0,0 +1,440 @@
+      expect(find.byType(RoundedRectanglePopoverAppearance), findsNothing);
+    });
+
+    testWidgetsOnAllPlatforms('enforces the given popover geometry', (tester) async {
```
If this is only testing height constraints, then please name the test accordingly. This name suggests that all aspects of geometry are being verified against all possible conditions.

Related to that, is testing just the height of the popover, in just one set of conditions, sufficient to lock down the geometry implementation?

angelosilvestre replied:
Updated.

matthew-carroll replied:
I'm still confused about why a single test is sufficient for this. Don't we have rules about where the popover appears based on surrounding space, and how large the popover is allowed to be? If so, then it seems like we should have a thorough collection of tests to lock down each of those behaviors...

angelosilvestre replied:
Ok, this test is to ensure we honor an app-configured geometry. We don't have tests for our default geometry. I'll add those.

matthew-carroll commented on super_editor/test/infrastructure/popover_test.dart:
```diff
@@ -0,0 +1,440 @@
+      expect(find.byType(RoundedRectanglePopoverAppearance), findsOneWidget);
+
+      // Ensure the maxHeight was honored.
+      expect(tester.getRect(find.byType(RoundedRectanglePopoverAppearance)).height, 10);
```
Please ensure that we're testing with reasonable sizes and shapes so that we don't let little things sneak by. I doubt that any real popover will be 10px tall.

matthew-carroll commented on super_editor/test/infrastructure/popover_test.dart:
```diff
@@ -0,0 +1,440 @@
+      expect(tester.getRect(find.byType(RoundedRectanglePopoverAppearance)).height, 10);
+    });
+
+    testWidgetsOnAllPlatforms('shares focus with SuperEditor', (tester) async {
```
This test seems overly specific. Do we actually care about the popover sharing focus with SuperEditor? Or do we care about sharing focus with any arbitrary widget that wants to share?

angelosilvestre replied:
Updated.

matthew-carroll commented on super_editor/test/infrastructure/popover_test.dart:
```diff
@@ -0,0 +1,440 @@
+    });
+  });
+
+  group('ItemSelectionList', () {
```
Should this group be in this test file?

matthew-carroll commented:
@angelosilvestre wanted to make sure you didn't forget about this PR.

angelosilvestre commented:
@matthew-carroll Updated the PR

angelosilvestre commented:
@matthew-carroll Added teste for the default popover aligner.

angelosilvestre commented on super_editor/test/infrastructure/popover_test.dart:
```diff
@@ -0,0 +1,493 @@
+          expect(popoverRect.top, greaterThan(buttonRect.bottom));
+        });
+
+        testWidgetsOnAllPlatforms('positions the popover above button if there is room', (tester) async {
```
Updated.

angelosilvestre commented on super_editor/test/infrastructure/popover_test.dart:
```diff
@@ -0,0 +1,493 @@
+          final popoverRect = tester.getRect(find.byType(RoundedRectanglePopoverAppearance));
+
+          // Ensure popover was pinned of the bottom to the boundary widget.
+          expect(popoverRect.bottom, 600);
```
Yeah, updated.

angelosilvestre commented on super_editor/test/infrastructure/popover_test.dart:
```diff
@@ -0,0 +1,493 @@
+        });
+
+        testWidgetsOnAllPlatforms(
+            'pins the popover to the bottom if there isn\'t room neither below or above the button', (tester) async {
```
I modified the tests to use a different screen size and different widget sizes with a comment explaining why.

angelosilvestre commented on super_editor/test/infrastructure/popover_test.dart:
```diff
@@ -0,0 +1,493 @@
+      });
+    });
+
+    testWidgetsOnAllPlatforms('shares focus with other widgets', (tester) async {
```
Updated.

matthew-carroll commented on super_editor/test/infrastructure/popover_test.dart:
```diff
@@ -0,0 +1,530 @@
+          expect(popoverRect.bottom, lessThan(buttonRect.top));
+        });
+
+        testWidgetsOnAllPlatforms('pins the popover to the bottom if there is not room below or above the button',
```
Can you clarify "to the bottom" of what? screen? boundary? button?

matthew-carroll commented:
In addition to earlier comments, I checked out this branch so I could work on paragraph alignment. I discovered that we're displaying too many alignment options in the drop down. This is because we're looping through all `TextAlign` values which means we're showing both `left` and `start`, and `right` and `end`. So we're essentially repeating equivalent options.

```dart
items: TextAlign.values
          .map(
            (alignment) => SuperEditorDemoIconItem(
              icon: _buildTextAlignIcon(alignment),
              id: alignment.name,
            ),
          )
          .toList(),
```

matthew-carroll commented:
@angelosilvestre I'm gonna merge this so that I can fix paragraph alignment, which I based off of this branch so that I could select the paragarph alignment from the toolbar. Please put up a followup PR with the final adjustments from the review.

matthew-carroll commented:
Also, please cherry pick this to stable, per usual.
